### PR TITLE
#243 BroadcastBuffer does not need `event` type for auto commit enabled

### PR DESCRIPTION
**Pull request by anilgursel**
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [x] Tests added.
- [x] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

This introduces an API change that's why it goes to 0.9.0.. To use `commit` for at least once semantics, one should prefer `PersistentBufferAtLeastOnce` and `BroadcastBufferAtLeastOnce`, these two will output `Event`.  `PersistentBuffer` and `BroadcastBuffer` will keep the input type.

### CODE REVIEW

File: squbs-pattern/src/main/scala/org/squbs/pattern/stream/BroadcastBufferBase.scala [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  private val outputPorts = queue.totalOutputPorts
+  private val in = Inlet[T]("BroadcastBuffer.in")
+  private val out = Vector.tabulate(outputPorts)(i ⇒ Outlet[S]("BroadcastBuffer.out" + i))
+  private val outWithIndex = out.zipWithIndex
+  val shape: UniformFanOutShape[T, S] = UniformFanOutShape(in, out: _*)
+  @volatile private var upstreamFailed = false
+  @volatile private var upstreamFinished = false
+  private val queueCloserActor = system.actorOf(Props(classOf[PersistentQueueCloserActor[T]], queue))
+
+  def elementOut(e: Event[T]): S = ???
```
Comment (akara) [inline squbs-pattern/src/main/scala/org/squbs/pattern/stream/BroadcastBufferBase.scala]: @anilgursel can you please take a look at this one?

File: squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentBufferBase.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+  private[stream] val in = Inlet[T]("PersistentBuffer.in")
+  private[stream] val out = Outlet[S]("PersistentBuffer.out")
+  val shape: FlowShape[T, S] = FlowShape.of(in, out)
+  val defaultOutputPort = 0
+  @volatile private var upstreamFailed = false
+  @volatile private var upstreamFinished = false
+  private val queueCloserActor = system.actorOf(Props(classOf[PersistentQueueCloserActor[T]], queue))
+
+  def elementOut(e: Event[T]): S = ???
```
Comment (akara) [inline squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentBufferBase.scala]: Ditto here.

File: squbs-pattern/src/test/scala/org/squbs/pattern/stream/BroadcastBufferAtLeastOnceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+    val head1 = Await.result(head1F, awaitMax)
+    val head2 = Await.result(head2F, awaitMax)
+    println(s"First record processed after shutdown => ${(format(head1.entry), format(head2.entry))}")
+    val last1 = Await.result(last1F, awaitMax)
+    val last2 = Await.result(last2F, awaitMax)
+    eventually { buffer.queue shouldBe 'closed }
+    assertions(beforeShutDown, SinkCounts(last1, last2), SinkCounts(totalProcessed, totalProcessed))
+  }
+
+  private def assertions(beforeShutDown: SinkCounts, afterRecovery: SinkCounts, totalRecords: SinkCounts) = {
```
Comment (akara) [inline squbs-pattern/src/test/scala/org/squbs/pattern/stream/BroadcastBufferAtLeastOnceSpec.scala]: I'm quite surprised the assertions for both the at least once use case and the no-commit use cases are the same. They provide different guarantees, right? This is for both persistent and broadcast buffer. Is our check leaking somewhere?

File: squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentBufferBase.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+              queueCloserActor ! PushedAndCommitted(defaultOutputPort, lastPushed, queue.read(defaultOutputPort))
+              queueCloserActor ! UpstreamFinished
+              completeStage()
+            } else downstreamWaiting = true
+        }
+      }
+    })
+  }
+
+  def commit[U] = Flow[Event[U]].map { element =>
```
Comment (anilgursel) [inline squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentBufferBase.scala]: I agree.  I will go with the second approach.

File: squbs-pattern/src/test/scala/org/squbs/pattern/stream/BroadcastBufferAtLeastOnceSpec.scala (hunk 3 above)
Comment (akara) [inline squbs-pattern/src/test/scala/org/squbs/pattern/stream/BroadcastBufferAtLeastOnceSpec.scala]: Thanks! Let me zoom into that a bit.

**Comment by akara** [comment]
@anilgursel LGTM. Please squash and I'll merge.

**Pull request merged**